File: Python/KodeProgram/GerlongElectric.py
# GerlongElectric.py
# Driver Program
# Program ini memanfaatkan inheritance multilevel dan polymorphism
# untuk mengelola data produk di toko elektronik.

from Laptop import Laptop
from Handphone import Handphone
from Aksesoris import Aksesoris

# Utility function: cetak tabel dinamis
def tampilkan_produk(daftar_produk):
    if not daftar_produk:
        print("Belum ada produk.")
        return

    headers = ["ID", "Nama", "Harga", "Brand", "Garansi", "Stok", "Kategori", "Detail"]

    # Hitung lebar kolom berdasarkan data terpanjang
    col_widths = [len(h) for h in headers]
    for p in daftar_produk:
        col_widths[0] = max(col_widths[0], len(str(p.get_id())))
        col_widths[1] = max(col_widths[1], len(p.get_nama()))
        col_widths[2] = max(col_widths[2], len(str(int(p.get_harga()))))
        col_widths[3] = max(col_widths[3], len(p.get_brand()))
        col_widths[4] = max(col_widths[4], len(str(p.get_garansi())))
        col_widths[5] = max(col_widths[5], len(str(p.get_stok())))
        col_widths[6] = max(col_widths[6], len(p.get_kategori()))
        col_widths[7] = max(col_widths[7], len(p.get_detail()))

    # Cetak garis pemisah
    def print_separator():
        for w in col_widths:
            print("-" * (w + 2) + "+", end="")
        print()

    # Cetak header
    print_separator()
    for i, h in enumerate(headers):
        print(f" {h:<{col_widths[i]}} |", end="")
    print()
    print_separator()

    # Cetak isi tabel
    for p in daftar_produk:
        print(f" {p.get_id():<{col_widths[0]}} |", end="")
        print(f" {p.get_nama():<{col_widths[1]}} |", end="")
        print(f" {int(p.get_harga()):<{col_widths[2]}} |", end="")
        print(f" {p.get_brand():<{col_widths[3]}} |", end="")
        print(f" {p.get_garansi():<{col_widths[4]}} |", end="")
        print(f" {p.get_stok():<{col_widths[5]}} |", end="")
        print(f" {p.get_kategori():<{col_widths[6]}} |", end="")
        print(f" {p.get_detail():<{col_widths[7]}} |", end="")
        print()
    print_separator()


# Tambah produk baru
def tambah_produk(daftar_produk, next_id):
    print("Pilih kategori:")
    print("1. Laptop\n2. Handphone\n3. Aksesoris")
    pilih = input("Pilihan: ")

    # Input atribut umum
    nama = input("Nama: ")
    harga = float(input("Harga: "))
    brand = input("Brand: ")
    garansi = int(input("Garansi (bulan): "))
    stok = int(input("Stok: "))

    if pilih == "1":
        ram = int(input("RAM (GB): "))
        storage = int(input("Storage (GB): "))
        processor = input("Processor: ")
        daftar_produk.append(Laptop(next_id[0], nama, harga, brand, garansi, stok, ram, storage, processor))
    elif pilih == "2":
        os = input("OS: ")
        camera = int(input("Camera (MP): "))
        battery = int(input("Battery (mAh): "))
        daftar_produk.append(Handphone(next_id[0], nama, harga, brand, garansi, stok, os, camera, battery))
    elif pilih == "3":
        jenis = input("Jenis: ")
        warna = input("Warna: ")
        compatibility = input("Compatibility: ")
        daftar_produk.append(Aksesoris(next_id[0], nama, harga, brand, garansi, stok, jenis, warna, compatibility))
    else:
        print("Pilihan salah.")
        return

    next_id[0] += 1
    print("Produk berhasil ditambahkan!")


# Main Program
def main():
    daftar_produk = []
    next_id = [1]

    # Produk default (5 item)
    daftar_produk.append(Laptop(next_id[0], "Laptop Gaming", 15000000, "ASUS", 24, 5, 16, 512, "i7")); next_id[0]+=1
    daftar_produk.append(Laptop(next_id[0], "Laptop Ultrabook", 12000000, "Dell", 18, 3, 8, 256, "i5")); next_id[0]+=1
    daftar_produk.append(Handphone(next_id[0], "Smartphone Pro", 8000000, "Samsung", 12, 10, "Android", 108, 5000)); next_id[0]+=1
    daftar_produk.append(Handphone(next_id[0], "iPhone 14", 15000000, "Apple", 12, 7, "iOS", 48, 4500)); next_id[0]+=1
    daftar_produk.append(Aksesoris(next_id[0], "Headphone BT", 1200000, "Sony", 6, 20, "Headset", "Hitam", "Bluetooth")); next_id[0]+=1

    while True:
        print("\n=== Menu Gerlong Electric ===")
        print("1. Tambah Produk")
        print("2. Lihat Semua Produk")
        print("3. Keluar")
        menu = input("Pilihan: ")

        if menu == "1":
            tambah_produk(daftar_produk, next_id)
        elif menu == "2":
            tampilkan_produk(daftar_produk)
        elif menu == "3":
            print("Keluar...")
            break
        else:
            print("Pilihan salah!")

if __name__ == "__main__":
    main()

File: Java/KodeProgram/GerlongElectric.java
// GerlongElectric.java
// File ini berisi main program untuk menampilkan menu interaktif.
import java.util.*;

public class GerlongElectric {
    private static Scanner sc = new Scanner(System.in);

    private static void printSeparator(int[] colWidths) {
        for (int width : colWidths) {
            for (int i = 0; i < width + 2; i++) System.out.print("-");
            System.out.print("+");
        }
        System.out.println();
    }

    private static void tampilkanProduk(List<Elektronik> daftarProduk) {
        if (daftarProduk.isEmpty()) {
            System.out.println("Belum ada produk yang ditambahkan.");
            return;
        }

        String[] headers = {"ID", "Nama", "Harga", "Brand", "Garansi", "Stok", "Kategori", "Detail"};
        int[] colWidths = new int[headers.length];

        for (int i = 0; i < headers.length; i++) {
            colWidths[i] = headers[i].length();
        }

        for (Elektronik p : daftarProduk) {
            colWidths[0] = Math.max(colWidths[0], String.valueOf(p.getId()).length());
            colWidths[1] = Math.max(colWidths[1], p.getNama().length());
            colWidths[2] = Math.max(colWidths[2], String.valueOf((long)p.getHarga()).length());
            colWidths[3] = Math.max(colWidths[3], p.getBrand().length());
            colWidths[4] = Math.max(colWidths[4], String.valueOf(p.getGaransi()).length());
            colWidths[5] = Math.max(colWidths[5], String.valueOf(p.getStok()).length());
            colWidths[6] = Math.max(colWidths[6], p.getKategori().length());
            colWidths[7] = Math.max(colWidths[7], p.getDetail().length());
        }

        printSeparator(colWidths);
        for (int i = 0; i < headers.length; i++) {
            System.out.printf(" %-" + colWidths[i] + "s |", headers[i]);
        }
        System.out.println();
        printSeparator(colWidths);

        for (Elektronik p : daftarProduk) {
            System.out.printf(" %-" + colWidths[0] + "s |", p.getId());
            System.out.printf(" %-" + colWidths[1] + "s |", p.getNama());
            System.out.printf(" %-" + colWidths[2] + "s |", (long)p.getHarga());
            System.out.printf(" %-" + colWidths[3] + "s |", p.getBrand());
            System.out.printf(" %-" + colWidths[4] + "s |", p.getGaransi());
            System.out.printf(" %-" + colWidths[5] + "s |", p.getStok());
            System.out.printf(" %-" + colWidths[6] + "s |", p.getKategori());
            System.out.printf(" %-" + colWidths[7] + "s |", p.getDetail());
            System.out.println();
        }
        printSeparator(colWidths);
    }

    private static void tambahProduk(List<Elektronik> daftarProduk, int[] nextId) {
        System.out.println("Pilih kategori:");
        System.out.println("1. Laptop\n2. Handphone\n3. Aksesoris");
        System.out.print("Pilihan: ");
        int pilih = sc.nextInt(); sc.nextLine();

        System.out.print("Nama: "); String nama = sc.nextLine();
        System.out.print("Harga: "); double harga = sc.nextDouble(); sc.nextLine();
        System.out.print("Brand: "); String brand = sc.nextLine();
        System.out.print("Garansi (bulan): "); int garansi = sc.nextInt();
        System.out.print("Stok: "); int stok = sc.nextInt(); sc.nextLine();

        if (pilih == 1) {
            System.out.print("RAM (GB): "); int ram = sc.nextInt();
            System.out.print("Storage (GB): "); int storage = sc.nextInt(); sc.nextLine();
            System.out.print("Processor: "); String proc = sc.nextLine();
            daftarProduk.add(new Laptop(nextId[0]++, nama, harga, brand, garansi, stok, ram, storage, proc));
        } else if (pilih == 2) {
            System.out.print("OS: "); String os = sc.nextLine();
            System.out.print("Camera (MP): "); int cam = sc.nextInt();
            System.out.print("Battery (mAh): "); int bat = sc.nextInt(); sc.nextLine();
            daftarProduk.add(new Handphone(nextId[0]++, nama, harga, brand, garansi, stok, os, cam, bat));
        } else if (pilih == 3) {
            System.out.print("Jenis: "); String jenis = sc.nextLine();
            System.out.print("Warna: "); String warna = sc.nextLine();
            System.out.print("Compatibility: "); String comp = sc.nextLine();
            daftarProduk.add(new Aksesoris(nextId[0]++, nama, harga, brand, garansi, stok, jenis, warna, comp));
        } else {
            System.out.println("Pilihan salah.");
        }
    }

    public static void main(String[] args) {
        List<Elektronik> daftarProduk = new ArrayList<>();
        int[] nextId = {1};
        int menu;

        daftarProduk.add(new Laptop(nextId[0]++, "Laptop Gaming", 15000000, "ASUS", 24, 5, 16, 512, "i7"));
        daftarProduk.add(new Laptop(nextId[0]++, "Laptop Ultrabook", 12000000, "Dell", 18, 3, 8, 256, "i5"));
        daftarProduk.add(new Handphone(nextId[0]++, "Smartphone Pro", 8000000, "Samsung", 12, 10, "Android", 108, 5000));
        daftarProduk.add(new Handphone(nextId[0]++, "iPhone 14", 15000000, "Apple", 12, 7, "iOS", 48, 4500));
        daftarProduk.add(new Aksesoris(nextId[0]++, "Headphone BT", 1200000, "Sony", 6, 20, "Headset", "Hitam", "Bluetooth"));

        do {
            System.out.println("\n=== Menu Gerlong Electric ===");
            System.out.println("1. Tambah Produk");
            System.out.println("2. Lihat Semua Produk");
            System.out.println("3. Keluar");
            System.out.print("Pilihan: ");
            menu = sc.nextInt(); sc.nextLine();

            if (menu == 1) tambahProduk(daftarProduk, nextId);
            else if (menu == 2) tampilkanProduk(daftarProduk);
            else if (menu == 3) System.out.println("Keluar...");
            else System.out.println("Pilihan salah!");
        } while (menu != 3);
    }
}